redis/src/main/scala/zio/redis/Interpreter.scala

```diff
@@ -1,135 +1,230 @@
+        op(channel)(completionHandlerCallback(k))
+        Left(IO.effect(channel.close()).ignore)
       }
 
-      private def unsafeSend(command: Chunk[String]): Unit = {
-        val data     = command.mkString
-        val envelope = s"*${command.length}\r\n$data"
-        val buffer   = ByteBuffer.wrap(envelope.getBytes(UTF_8))
+    def socket(host: String, port: Int): Layer[IOException, ByteStream] =
+      socket(IO.effectTotal(new InetSocketAddress(host, port)))
 
-        while (buffer.hasRemaining())
-          channel.write(buffer)
+    def socket(address: SocketAddress): Layer[IOException, ByteStream] = socket(UIO.succeed(address))
+
+    def socketLoopback(port: Int): Layer[IOException, ByteStream] =
+      socket(IO.effectTotal(new InetSocketAddress(InetAddress.getLoopbackAddress, port)))
+
+    private def socket(getAddress: UIO[SocketAddress]): Layer[IOException, ByteStream] = {
+      val makeBuffer = IO.effectTotal(ByteBuffer.allocateDirect(ResponseBufferSize))
+      ZLayer.fromEffect {
+        for {
+          address     <- getAddress
+          readBuffer  <- makeBuffer
+          writeBuffer <- makeBuffer
+        } yield new Connection(address, readBuffer, writeBuffer)
       }
     }
+
+    private final class Connection(
+      address: SocketAddress,
+      readBuffer: ByteBuffer,
+      writeBuffer: ByteBuffer
+    ) extends Service {
+
+      private def openChannel: Managed[IOException, AsynchronousSocketChannel] = {
+        val make = for {
+          channel <- IO.effect {
+                       val channel = AsynchronousSocketChannel.open()
+                       channel.setOption(StandardSocketOptions.SO_KEEPALIVE, Boolean.box(true))
+                       channel.setOption(StandardSocketOptions.TCP_NODELAY, Boolean.box(true))
+                       channel
+                     }
+          _       <- effectAsyncChannel[AsynchronousSocketChannel, Void](channel)(c => c.connect(address, null, _))
+        } yield channel
+        Managed.fromAutoCloseable(make).refineToOrDie[IOException]
+      }
+
+      override val connect: Managed[IOException, ReadWriteBytes] =
+        openChannel.map { channel =>
+          val readChunk =
+            (for {
+              _     <- IO.effectTotal(readBuffer.clear())
+              _     <- effectAsyncChannel[AsynchronousByteChannel, Integer](channel)(c => c.read(readBuffer, null, _))
+                     .filterOrFail(_ >= 0)(new EOFException())
+              chunk <- IO.effectTotal {
+                         readBuffer.flip()
+                         val count = readBuffer.remaining()
+                         val array = Array.ofDim[Byte](count)
+                         readBuffer.get(array)
+                         Chunk.fromArray(array)
+                       }
+            } yield chunk).mapError {
+              case _: EOFException => None
+              case e: IOException  => Some(e)
+            }
+
+          @inline
```

Comment:
Hehe, I meant to inline the function completely, but annotation can work as well.

Reply:
Oh I see, but it's recursive, so we can't.